thirdparty/DiskANN/include/pq_flash_index.h

```diff
@@ -195,6 +196,7 @@ namespace diskann {
 
     std::string                        disk_index_file;
     std::vector<std::pair<_u32, _u32>> node_visit_counter;
+    _u32                               search_counter = 0;
```

Comment:
CMIIW: the warm up method performs`sample_num` searches and cache the most visited `num_nodes_to_cache` nodes. If user requested searches come in during the time, those are also counted towards `sample_num` and will affect what node to add to `nodes_to_cache`. **The searches performed by the warm up method are all in a single thread**, but to protect from concurrent access by user searches from other threads, we used several atomic primitives to protect `search_counter, node_visit_counter` and others.

I wonder do we have to count user searches during the warm up? If not we can greatly simplify the code. Since the warm up searches are serial we could just count them without using any atomics. 

```cpp
template <warm_up = false>
cached_beam_search(q) {
  if constexpr (warm_up) {
    // no need to protect node_visit_counter
    count_node_visit();
  }
}

void generate_cache_list_from_sample_queries() {
  // single thread operation:
  // perform all `sample_num` searches, not counting user searches
  for (q : warm_up_qs) {
    cached_beam_search<true>(q);
  }
  top_nodes = sorted(node_visit_counter)[:num_nodes_to_cache];
  load_into_cache(top_nodes);
}

void normal_search(q) {
  // multiple threads may enter here at the same time
  // user searches don't affect `node_visit_counter`
  cached_beam_search(q);
}
```

The current code is not completely thread safe anyway, when we have finished warm up, before we flip the bit `reinterpret_cast<std::atomic<bool> &>(this->count_visited_nodes).exchange(false)`, other threads may continue updating `search_counter` and `node_visit_counter` when sorting `node_visit_counter`.